src/exchangeRateRouter.ts

```diff
@@ -329,62 +330,110 @@ const getRedisMarkets = async (
-      const rate = Number(exchangeRate)
+    // Only parse if jsonString is not null/undefined
+    let redisResult: CoinrankRedis | undefined
+    if (jsonString !== null && jsonString !== undefined) {
+      try {
+        redisResult = JSON.parse(jsonString)
+
+        // If we have valid data that hasn't expired, return it
+        const lastUpdated =
+          redisResult != null ? new Date(redisResult.lastUpdate).getTime() : 0
+        if (nowTimestamp - lastUpdated <= EXPIRE_TIME) {
+          return redisResult
+        }
+        // We have cached data but it's expired - we'll try to refresh it below
+        // but will fall back to this expired data if refresh fails
+      } catch (e) {
+        // JSON parsing error, redisResult remains undefined
+        logger(`Error parsing Redis data for ${fiatCode}: ${e}`)
+        // Continue to try to get fresh data
+      }
+    }
 
-      // Get USD rankings
-      const jsonString = await getAsync(
-        `${REDIS_COINRANK_KEY_PREFIX}_${defaultFiatCode}`
-      )
-      redisResult = JSON.parse(jsonString)
-      let { markets } = redisResult
-
-      // Modify fiat-related fields with the forex rate
-      markets = markets.map(m => ({
-        ...m,
-        marketCap: m.marketCap * rate,
-        price: m.price * rate,
-        volume24h: m.volume24h * rate,
-        high24h: m.high24h * rate,
-        low24h: m.low24h * rate,
-        priceChange24h: m.priceChange24h * rate,
-        marketCapChange24h: m.marketCapChange24h * rate,
-        circulatingSupply: m.circulatingSupply * rate,
-        totalSupply: m.totalSupply * rate,
-        maxSupply: m.maxSupply * rate,
-        allTimeHigh: m.allTimeHigh * rate,
-        allTimeLow: m.allTimeLow * rate
-      }))
-
-      // Update redis cache
-      const redisData: CoinrankRedis = {
-        markets,
-        lastUpdate: now.toISOString()
+    // If we need to convert from USD (either no data or expired data)
+    if (fiatCode !== defaultFiatCode) {
+      try {
+        // Get exchange rate
+        const result = await fetch(
+          `${ratesServerAddress}/v2/exchangeRate?currency_pair=${defaultFiatCode}_${fiatCode}`
+        )
+        if (!result.ok) {
+          throw new Error(`Exchange rate API returned status ${result.status}`)
+        }
+
+        const resultJson = await result.json()
+        const { exchangeRate } = asExchangeRateResponse(resultJson)
+        const rate = Number(exchangeRate)
+
+        // Validate the rate
+        if (rate == null || isNaN(rate) || rate <= 0) {
+          throw new Error(`Invalid exchange rate: ${exchangeRate}`)
+        }
+
+        // Get USD rankings
+        const usdJsonString = await getAsync(
+          `${REDIS_COINRANK_KEY_PREFIX}_${defaultFiatCode}`
+        )
+        if (usdJsonString == null) {
+          throw new Error(`No USD data available in Redis`)
+        }
+
+        const usdRedisResult = JSON.parse(usdJsonString)
+        let { markets } = usdRedisResult
```

Comment:
Nitpick: make this a const and create a new non-defaultfiat const when calling .map instead